sirepo/auth_role_moderation.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+u"""Moderate user roles
+
+:copyright: Copyright (c) 2018-2019 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+import flask
+from pykern import pkconfig
+from pykern.pkdebug import pkdp, pkdlog
+from pykern.pkcollections import PKDict
+from pykern import pkjinja
+import sirepo.api_perm
+import sirepo.auth_db
+import sirepo.http_reply
+import sirepo.http_request
+import sirepo.simulation_db
+
+_cfg = None
+
+@sirepo.api_perm.require_user
+def api_saveModerationReason():
+    def _moderation_link(sim_type):
+        return flask.url_for('_dispatch_empty', _external=True).rstrip('/') \
+               + sirepo.uri.local_route(sim_type, 'admRoles')
+
+    def _dns_reverse_lookup(ip):
+        import dns.resolver, dns.reversename
+        try:
+            return str(dns.resolver.query(dns.reversename.from_address(ip), 'PTR')[0])
+        except (dns.resolver.NoAnswer, dns.resolver.NXDOMAIN):
+            return 'No Reverse DNS Lookup'
+
+    def _send_moderation_email(info):
+        sirepo.smtp.send(
+            recipient=_cfg.moderator_email,
+            subject=f'{info.sim_type} Access Request',
+            body=pkjinja.render_resource('auth_role_moderation/moderation_email', info),
+        )
+
+    req = sirepo.http_request.parse_post(require_sim_type=False)
+    d = req.req_data
+    uid = sirepo.auth.logged_in_user()
+
+    sirepo.auth_db.UserRoleInvite(
```

Comment:
If I user enters this information twice this may fail with an IntegrityError. I think we should catch that error, log it, and then return a UserAlert to the user letting them know that they've already supplied info and we are processing it.

Reply:
f77e9717949945b988991128a2701fe97b1111d6